java/core/lance-jni/src/lib.rs

```diff
@@ -88,3 +99,174 @@ pub extern "system" fn Java_com_lancedb_lance_JniLoader_initLanceLogger() {
+            builder_ptr as jlong
+        }
+        Err(e) => {
+            env.throw_new("java/lang/RuntimeException", &e.to_string()).unwrap();
+            0
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_whenMatchedUpdateAllNative(
+    mut env: JNIEnv,
+    _class: JClass,
+    builder_handle: jlong,
+    condition: JString,
+) {
+    let builder = unsafe { &mut *(builder_handle as *mut LanceMergeInsertBuilder) };
+
+    let when_matched = if condition.is_null() {
+        WhenMatched::UpdateAll
+    } else {
+        let condition_str: String = env.get_string(&condition).unwrap().into();
+        match WhenMatched::update_if(&builder.dataset, &condition_str) {
+            Ok(matched) => matched,
+            Err(e) => {
+                env.throw_new("java/lang/RuntimeException", &e.to_string()).unwrap();
+                return;
+            }
+        }
+    };
+
+    builder.when_matched(when_matched);
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_whenNotMatchedInsertAllNative(
+    _env: JNIEnv,
+    _class: JClass,
+    builder_handle: jlong,
+) {
+    let builder = unsafe { &mut *(builder_handle as *mut LanceMergeInsertBuilder) };
+    builder.when_not_matched(WhenNotMatched::InsertAll);
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_whenNotMatchedBySourceDeleteNative(
+    mut env: JNIEnv,
+    _class: JClass,
+    builder_handle: jlong,
+    expr: JString,
+) {
+    let builder = unsafe { &mut *(builder_handle as *mut LanceMergeInsertBuilder) };
+
+    let when_not_matched_by_source = if expr.is_null() {
+        WhenNotMatchedBySource::Delete
+    } else {
+        let expr_str: String = env.get_string(&expr).unwrap().into();
+        match WhenNotMatchedBySource::delete_if(&builder.dataset, &expr_str) {
+            Ok(not_matched) => not_matched,
+            Err(e) => {
+                env.throw_new("java/lang/RuntimeException", &e.to_string()).unwrap();
+                return;
+            }
+        }
+    };
+
+    builder.when_not_matched_by_source(when_not_matched_by_source);
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_conflictRetriesNative(
+    _env: JNIEnv,
+    _class: JClass,
+    builder_handle: jlong,
+    max_retries: jint,
+) {
+    let builder = unsafe { &mut *(builder_handle as *mut LanceMergeInsertBuilder) };
+    builder.conflict_retries(max_retries as u32);
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_retryTimeoutNative(
+    _env: JNIEnv,
+    _class: JClass,
+    builder_handle: jlong,
+    timeout_millis: jlong,
+) {
+    let builder = unsafe { &mut *(builder_handle as *mut LanceMergeInsertBuilder) };
+    builder.retry_timeout(Duration::from_millis(timeout_millis as u64));
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_MergeInsertBuilder_executeNative(
```

Comment:
we should try to consolidate this with what is already there in `BlockingDataset` and make `merge_insert` another method of it.